src/s390x/fba.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::blockdev::get_sector_size;
+use crate::errors::*;
+use crate::s390x::dasd::{partitions_from_header, Dasd, Range};
+use mbrman::{MBRPartitionEntry, CHS, MBR};
+use rand::Rng;
+use std::fs::File;
+
+pub struct FbaDasd {
+    pub(crate) dasd: String,
+}
+
+impl FbaDasd {
+    pub fn new<S: AsRef<str>>(dasd: S) -> FbaDasd {
+        FbaDasd {
+            dasd: dasd.as_ref().to_string(),
+        }
+    }
+}
+
+impl Dasd for FbaDasd {
+    fn prepare(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn make_partitions(&self, device: &mut File, first_mb: &[u8]) -> Result<Vec<Range>> {
+        let bytes_per_block = get_sector_size(&device).unwrap().get().into();
+        let partitions = partitions_from_header(bytes_per_block, first_mb)?;
+        let id: u32 = rand::thread_rng().gen();
+        let mut ranges = Vec::new();
+        let mut mbr = MBR::new_from(device, bytes_per_block as u32, id.to_be_bytes())
+            .chain_err(|| format!("creating new partition table on {}", self.dasd))?;
+
+        for (idx, it) in partitions.iter().enumerate() {
+            let size = it.ending_lba - it.starting_lba + 1;
+            let offset = it.starting_lba * bytes_per_block;
+            ranges.push(Range {
+                in_offset: offset,
+                out_offset: offset,
+                length: size * bytes_per_block,
+            });
+            mbr[idx + 1] = MBRPartitionEntry {
+                boot: false,
+                first_chs: CHS::empty(),
+                sys: 0x83, // MBR_LINUX_DATA_PARTITION
+                last_chs: CHS::empty(),
+                starting_lba: it.starting_lba as u32,
+                sectors: size as u32,
```

Comment:
Let's use `try_into()` on these two lines to catch any overflow from a malformed image.